File: Leetcode/993_Cousins_in_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* xTreeNode;
    int xDepth;
    TreeNode* yTreeNode;
    int yDepth;
    bool isCousins(TreeNode* root, int x, int y) {
        helper(root, x, y, 0, root);
        
        if(xDepth != yDepth)
        {
            return false;
        }
        
        if(xTreeNode == yTreeNode)
        {
            return false;
        }
        
        return true;
    }
    
    void helper(TreeNode* root, int x, int y, int depth, TreeNode* parent)
    {
        if(root == nullptr)
        {
            return;
        }
        if(root->val == x)
        {
            xTreeNode = parent;
            xDepth = depth;
        }
        else if(root->val == y)
        {
            yTreeNode = parent;
            yDepth = depth;
        }
        helper(root->left, x, y, depth + 1, root);
        helper(root->right, x, y, depth + 1, root);
    }
};

File: Leetcode/0993/993_Cousins_in_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
let leftDepth = 0;

var isCousins = function (root, x, y) {
    helper(root, x, 0);
    return traversal(root, y, 0, root, x);
};

var helper = function (root, target, depth) {
    if (root) {
        helper(root.left, target, depth + 1);
        if (root.val === target) {
            leftDepth = depth;
            return;
        }
        helper(root.right, target, depth + 1);
    }
}

var traversal = function (root, target, depth, parentNode, compareValue) {
    if (root && depth <= leftDepth) {
        if (root.val === target && depth === leftDepth) {
            if ((parentNode.left && parentNode.left.val === compareValue) || (parentNode.right && parentNode.right.val === compareValue)) {
                return false;
            }
            return true;
        }
        return traversal(root.right, target, depth + 1, root, compareValue) || traversal(root.left, target, depth + 1, root, compareValue);
    }
    return false;
}